### ISSUE TITLE
[Feature Request] Open Food Fact Integration

### ISSUE THREAD

**Issue by HydrelioxGitHub** [grocy/grocy#158]
It could be very time consuming to input data when you create your database of product. There is an open database called OpenFooFact with a lot of data on every product we eat. (Maybe you already know about this database).

The usecase could be this : 
If I want to add a new product, I can only put the barcode number and the product could be created automatically and populated with info from OpenFoodFact.

The link to the database info : [https://fr.openfoodfacts.org/](https://fr.openfoodfacts.org/)
The link to the API documentation :[https://en.wiki.openfoodfacts.org/API)]( https://en.wiki.openfoodfacts.org/API)

**Comment by berrnd** [comment]
This has been discussed before (#6).

> It could be very time consuming to input data when you create your database of product.

I disagree. I use grocy since almost 2 years on daily basis. When most of the products are added, I think you will rarely add new ones. Also, do you really want to have each and every product as a different product in grocy? For example I personally only have "Chocolate (Milk)" and barcodes from different brands are attached to this product. So regardless which one I buy/scan, it is always resolved to "my" product. Additionally, when I buy/scan a new brand, the barcode can be attached to the already existing product directly during the purchase workflow (a popup when you enter something unknown in the product dropdown field). The same applies for creating products, this can also be done directly during the booking of a purchase.

Anyway, through the API it's already possible to add new products by external services, see [here](https://github.com/grocy/grocy#barcode-lookup-via-external-services).

So I still don't really see the benefits of external barcode lookups, maybe we can discuss about how to integrate the current "API way" in the UI.

**Comment by HydrelioxGitHub** [comment]
Thank you for your answer. 
I agree with you that when all your products are added there is no really need for that. But if you have to add tens of products for starting Grocy it could be useful.
I already agree with the fact that it's good to have one generic product with many brands associated with. 
I'm pretty new to Grocy, and this feature request seems to be good at first sight but maybe need more thinking.

Thanks for your time anyway.

**Comment by berrnd** [comment]
No problem, all ideas are very welcome. :)

I also find this interesting in general, so I consider to do the following two things:
- Integrate the "barcode plugin lookup API only thing" into the UI
- Create a plugin for openfoodfacts.org

Afterwards, maybe this would be useful then:
- Display additional information provided by OFF in the product card
- When a scanned barcode cannot be resolved, query the API/Plugin and do a fuzzy search via the product name to automagically map it to a already existing generic product in grocy

If you have further detailed ideas, feel free to add them here.

**Comment by Altheran88** [comment]
Here is my take on a full integration of OpenFoddFacts

Barcode scanning app for Grocy (Other issue)
Country specified
Language Specified
Api to loggin to Openfoodfacts account.

Grocy Server : 

Country specified
Language Specified
Api to loggin to Openfoodfacts account.

Item Data : 
Picture
Common Name (OpenFood / Local)
Quantity
Brand
Link to product page
Foreign table : Categories (OpenFood / Local)
Foreign Table : Stores where sold and checkbox to specify stores where you buy stuff
Foreign table : Countries Sold
Nutrients
Ingredients (Good for allergies, you could specify allergies in your household)

Cellphone App :
Api to authenticate to OpenFoodfact and allow upload of data (contribute Data)
Api to authenticate on local server.
Option to contribute Data (When you scan an item, if incomplete or missing on OFF, will ask if you want to upload missing data to openfoodfacts)
Barcode scan -> Lookup local data and OFF.
Local : If not present, will download data from OFF and do action, else do action.
OFF : If Data missing and contributing, confirm upload of missing info, else do nothing.

Sync jobs or live update on server to upload/download missing informations of items.

**Comment by BlueBlueBlob** [comment]
Hi !
I'm trying to do simple integration between Grocy and openfood/product facts. 
But I'm running Grocy through hass.io add-on.
For now my solution is manual because I can't load the full db on a raspberry pi and I want it works offline.
For now I scan all my barcodes, do a select in the db dump (on a PC). Extract to influxdb format (handled by home-assistant).
So I can automate input barcode, grab info's for product and put in Grocy with Grocy api :)
(Work in progress)
Just to say it can be done 👍

**Comment by RaitoBezarius** [comment]
It appears to me that the current solution (writing an external Open Food Facts barcode resolver plugin) is not enough because it does not integrate with the integrated barcode scanner, does it?
Ideally, I see two solutions: take a DB dump of Open Food Facts and load it into Grocy through some script or add a way in Grocy to perform adhoc actions when a barcode is not known during scan.

Like, if I try to add by barcode a certain product and I enabled ad-hoc resolution, Grocy will go through all current and enabled plugins and once one of them returns a result, I take it. Or I could show all the results and let the user choose the most relevant one. This way, it opens up multiple barcode resolvers (not only Open Food Facts).

And we can easily bake Open Food Facts resolver as an optional resolver, IMHO, that's super cool because if you can have Internet on your server:

(1) No need to store the whole DB on your server

(2) Always up-to-date

(3) Your (non-custom) barcodes work out of the box

And we can add some nice features like NutriScore computations, vegetarian/vegan status, etc.

I could send a PR if you think that it'd be a good idea to add optional ad-hoc resolution during a barcode scan (server-side).

IMHO, it is just an HTTPS request which returns some JSON. It's not super expensive in terms of performance, and we can definitely add some caching if that's required (but we should just store the barcode definitely I suppose).

**Comment by berrnd** [comment]
Any contributions are welcome.

But it would be also required to do some mapping if the external lookup returns an already existing product for an unknown barcode (by product name?) and map the quantity units to the grocy ones. If that does not happen, you will have a new product with the name prefilled - still don't see the benefit about only that...

**Comment by BlueBlueBlob** [comment]
Hi all.
For my case, I just made script, to get all barcodes, download pic from OFF and upload them in Grocy.

Running once is enough or on a new product. 
For my case, I don't need more.

Good luck with the plugging 👍

**Comment by beetle442002** [comment]
@berrnd I have a few questions related to stock and this topic.

- Am I correct in thinking that the stock table is used for current inventory and that when a item is used that record/row is removed from the table?

- If that is the case does the row representing that specific item remain in the stock_log table for historic and analytic purposes?

- What is the parent product on the new product page?

**Comment by berrnd** [comment]
> - Am I correct in thinking that the stock table is used for current inventory and that when a item is used that record/row is removed from the table?

Yes, but there are `n` rows per product, they are just created as they came during purchase or inventory, on use a row will maybe get splitted ([code ref](https://github.com/grocy/grocy/blob/8da13ab22b43e9ca26fd8410e55018abf3e2ac31/services/StockService.php#L275-392)).

> - If that is the case does the row representing that specific item remain in the stock_log table for historic and analytic purposes?

Yes, just a log of all actions, displayed on a [`/stockjournal`](https://demo-prerelease.grocy.info/stockjournal).

> - What is the parent product on the new product page?

It's just to define parent-/sub products. E. g. "Chocolate" as a parent product and "Milk Chocolate" and "Dark Chocolate" as sub products. Snippet from the changelog of v2.5.0:

> - _Products can now have variations (nested products)_
>    - _Define the parent product for a product on the product edit page (only one level is possible, means a product which is used as a parent product in another product, cannot have a parent product itself)_
>    - _Parent and sub products can have stock (both are regular products, no difference from that side)_
>    - _On the stock overview page, the aggregated amount is displayed next to the amount (sigma sign)_
>    - _When a recipe needs a parent product, the need is also fulfilled when enough sub product(s) are in stock_

**Comment by bbreton09** [comment]
> Hi all.
> For my case, I just made script, to get all barcodes, download pic from OFF and upload them in Grocy.
> 
> Running once is enough or on a new product.
> For my case, I don't need more.
> 
> Good luck with the plugging 👍

Hi @BlueBlueBlob , your script is visible somewhere ?
Thank's

**Comment by teolemon** [comment]
Hello :-) Pierre, from Open Food Facts
We'd love to see more cross integration with OFF.
For those interested, we have a chatroom : https://slack.openfoodfacts.org (make sure to introduce yourselves).
We also have an official mobile app to scan and contribute new products, and we have a very basic list system (it might be awesome to add a Grocy sync plugin)

**Comment by PhilipWhiteside** [comment]
@berrnd 

> Also, do you really want to have each and every product as a different product in grocy? 

If you group them all into one, then the kcal count is not correct, and price. If you group them into a parent product, you can still select that generic product for a recipe, but consume the child product. There is a request to have that child selectable #605 too.

Sure I don't switch shops too much, but there is a local one I walk to, and another one I order online from. We will try different places now and then, where this could help. 

I think the biggest benefit of this is for new users.

**Comment by hype-armor** [comment]
> > Hi all.
> > For my case, I just made script, to get all barcodes, download pic from OFF and upload them in Grocy.
> > Running once is enough or on a new product.
> > For my case, I don't need more.
> > Good luck with the plugging 👍
> 
> Hi @BlueBlueBlob , your script is visible somewhere ?
> Thank's

this might be it.
https://github.com/Forceu/barcodebuddy

**Comment by BlueBlueBlob** [comment]
> > Hi all.
> > For my case, I just made script, to get all barcodes, download pic from OFF and upload them in Grocy.
> > Running once is enough or on a new product.
> > For my case, I don't need more.
> > Good luck with the plugging +1
> 
> Hi @BlueBlueBlob , your script is visible somewhere ?
> Thank's

https://github.com/BlueBlueBlob/appdaemon_scripts

**Comment by BlueBlueBlob** [comment]
In fact, just downloading pictures is not so hard.
But have a nice view where you can contribute (because it's the goal of OFF) is not my skills.

And If your script just downloads pics, plz check pic update.

Regards

**Comment by dvankley** [comment]
I'm reasonably sure [Barcode Buddy](https://github.com/Forceu/barcodebuddy) integrates with OFF and Grocy.

**Comment by basxto** [comment]
I’m trying to build a [little python program](https://github.com/basxto/grofff) to scrape data from OFF based on barcodes and update it, but the updating part is missing so far:

```
$ ./grofff.py -i 64
Product #64:
 Calories based on 344 kcal per 100g
 Name:
  Current value: Edeka Bio Rote Linsen
  Proposed value [a]: Rote Linsen geölt
  Proposed value [b]: Edeka Bio Rote Linsen geölt
 Total [c]alories:
  Current value: 0 kcal
  Proposed value: 1720 kcal
 One stock quantity is equal to 500 of quantity 5 (g) (not supported by grocy API)
y/N/abcq: bc
Updating products is not yet implemented:
 Name: Edeka Bio Rote Linsen geölt
 Total calories: 1720
```
I map OFF packaging strings to quantity IDs.
Getting products by barcode only work for products in stock, since the API is limited in that way.

The data on OFF does not seem very standardized, but good enough to get names, calories and purchase quantities.
Names sometimes contain the brand, sometimes not.
Packaging types also seem very much all over the place.

And the best thing, I can do for missing data on OpenFoodFacts, so far is to warn.

```
$ ./grofff.py -i 14
Product #14:
 Calories missing on OpenFoodFacts
 Name:
  Current value: Mio Mio Mate + Ginger
  Proposed value [a]: Mio Mio Mate Ginger
  Proposed value [b]: Ginger,Mio Mio Mate Mio Mio Mate Ginger
 Purchase [q]uantity
  Current value: 2 (Stück)
  Proposed value: 11 (Glas)
y/N/abcq: y
Updating products is not yet implemented:
 Name: Mio Mio Mate Ginger
 Purchase quantity: 11
```

And yes, I basically have one entry per product and use glas/bottle/package etc. as stock quantities, which I convert to SI units with custom quantity conversions.

**Comment by berrnd** [comment]
>  Getting products by barcode only work for products in stock, since the API is limited in that way.

That's not true, the endpoint `/stock/products/by-barcode/{barcode}` also returns the product if not in stock.
Or use the `/objects/{entity}` endpoint and iterate over all products yourself.
Or use (in the next version) the extended filters for (nearly) any endpoint added by #985.

**Comment by basxto** [comment]
> the endpoint `/stock/products/by-barcode/{barcode}` also returns the product if not in stock.
Oh, I assumed that since it’s `/stock/`.
Then it should work, that’s what I’m using.

And updating products works now, too.

**Comment by trevor87** [comment]
I also wanted this feature and played around a bit and came up with the following greasemonkey script: https://github.com/trevor87/grocy_autofood

It doesn't help that much but it might be a starting point if someone wants to tinker with it.

**Comment by CloCkWeRX** [comment]
This might be closable, with https://github.com/Forceu/barcodebuddy having an openfoodfacts backend?

**Comment by berrnd** [comment]
> This might be closable, with https://github.com/Forceu/barcodebuddy having an openfoodfacts backend?

No. This is a feature request to integrate OFF directly into grocy. Barcode Buddy is a community contributed external application (which works great). But completely unrelated to the topic here.

**Comment by luckydonald** [comment]
Especially for new users it would be an enormous benefit!  
Yes, if you've been using this for ages, sure you'll have everything you need already in your database and you can decide to never buy something unknown.
But I think for something so easy to make (and people already wanted to step up and do a PR), it's a bummer we don't already have this fallback mechanism for unknown barcodes.


I think having that [strict requirement](#issuecomment-568688509) for also finding duplicates is a bit of a blocker for this, and should be a separate feature at a later point in time.
For now it could be solved with simply asking the user if it should be added as a new item, e.g. "_Could not find that barcode. Do you want to search it online and add this as a new product?_ [Yes]  [No]".
Worst case, you have to delete something which was duplicated. That's a click. Having to enter new product information is so much more work than removing a duplicate.

**Comment by luckydonald** [comment]
Because
> [...] Barcode Buddy is a community contributed external application (which works great). But completely unrelated to the topic here.

As such it won't work with the iOS app, etc.

**Comment by rkantos** [comment]
> > As such it won't work with the iOS app
> 
> Having this supported by any of the great community contributed apps is just another different topic, not tracked here.
> 
> I personally use grocy daily on Android, without an app resp. the app is called "Firefox". As long as iOS has a webbrowser, grocy and Barcode Buddy will work also there.
> 
> Since Barcode Buddy provides all and would be kind of obsolete when implementing everything in grocy itself, it's just not on my list of the things to do next for grocy. I know, unpopular opinion, among so much others.
> 
> And fun fact: [Grocy Android](https://github.com/patzly/grocy-android) also already has an OFF integration since v2.0.0.

The Grocy-Android indeed has great functionality with it's MLKit barcode scanner and Open Food Fact query/scraping. I noticed this functionality a couple weeks back and decided to add support to Grocy for the grocery store that I use the most. You can get it from here and modify for your own shop of course: https://github.com/rkantos/grocy-android/commit/c0131e53bdd0bfc2b6672a597ac11778a9a0d1b1

You only need to edit the URL here: https://github.com/rkantos/grocy-android/blob/master/app/src/main/java/xyz/zedler/patrick/grocy/api/FoodieApi.java#L41

and the css-selector for product name here: https://github.com/rkantos/grocy-android/blob/master/app/src/main/java/xyz/zedler/patrick/grocy/helper/DownloadHelper.java#L2808

I didn't fully integrate it with Grocy-Android as how OFF (tooltips etc) is done as I didn't plan to PR it. Instead I made a request to Grocy-Android for a more generic feature as I see it could very easily be added to it (might even do it myself): https://github.com/patzly/grocy-android/issues/556

**Comment by bnordgren** [comment]
>... Also, do you really want to have each and every product as a different product in grocy? For example I personally only have "Chocolate (Milk)" and barcodes from different brands are attached to this product. So regardless which one I buy/scan, it is always resolved to "my" product. 

An interesting aspect of a true OpenFoodFact (OFF) integration might be the utilization of the ["Categories" taxonomy](https://github.com/openfoodfacts/openfoodfacts-server/blob/main/taxonomies/food/categories.txt) and its subsequent application to various Products. For instance if your "Product" was actually a category as listed on OFF, you might be able to list "All purpose flour" as a Product in your virtual kitchen, as well as referenced in recipes, while the external barcode lookup resolves any specific item within that group to the category/Product.  This seems to be implementing the strategy you describe, but the "groups" (taxonomy) have been vetted by a community, some categories have references to wikidata items, and the mapping from products to categories has already been done.

It does seem that for recipes you probably don't often care which brand of flour you purchase, but for other aspects of Grocy (e.g. cost tracking) you really do want to know specifically what was purchased, when, where and for how much. 

Alternatively, if you do represent each specific brand item belonging to a category of things as its own Product, Products might be able to have a field called "Category" which could be either a string or an array of strings. When determining whether you have stock on hand or not for a recipe, the ingredients list of the recipe could be matched against the category field of the product. If Grocy has no knowledge of the taxonomy, this will work for a single level. If you wanted to check the named category as well as sub- and sub-sub- categories, then you may have to perform an API call against OFF, or import the taxonomy into a reasoning engine or something.

So I guess there's two ways this could go: 
1. Keep the relationship to Recipe Ingredients<->Products the way it is, Products represent categories in a taxonomy and let the barcode lookup match products based on categories assigned to the looked-up-item; or
2. Change the relationship of Recipe Ingredients<->Products to include matches against a "Category" field, which is either populated automatically from an external source like OFF or manually.

Cast in the light of a full categorization taxonomy, it seems like Parent Products may cover membership of Products in the very lowest level (most specific) category and Product Groups would cover the very top level (the most general) categories, and there's a disconnect because all the middle layers are missing.

Huh. Thinking about it, the "disconnect" could be eliminated under scenario 2 if the "Category" field were multi-valued and the Category field named not only the Category it directly belongs to, but all the parents of that category up to the top level. You may not need to import someone else's taxonomy or leverage a reasoning engine.

**Comment by bnordgren** [comment]
@berrnd Just out of curiousity, I'm wondering what's the perception of "on-topic" for this ticket? If the solution to the OP (and other's problem) is to use barcode buddy, I'd kind of expect the ticket to be closed, as the claim is that the problem is solved. Yet it seems this has already been proposed and rejected in https://github.com/grocy/grocy/issues/158#issuecomment-468812212. But then two posts down from this claim that the ticket should stay open: "[Barcode Buddy](https://github.com/Forceu/barcodebuddy) already provides all (and more) this is about. Why not just use it, if that's such a big problem for anyone," which seems to communicate the design intent that integration with a specific external database is the job of external plugins and not the core software product. I'm getting mixed signals.

There's a list above of potential work items above, but it is unclear which are in scope or which have been completed or are holding the ticket open. Of the tasks in https://github.com/grocy/grocy/issues/158#issuecomment-468812212, I'm guessing (from the outside looking in) that the Barcode lookup API is done, and the provision of a "Grocy-native" barcode buddy clone is the hold out? Alternatively, there's also a vision of a [comprehensive OFF integration](https://github.com/grocy/grocy/issues/158#issuecomment-476576212) above, which is basically a full fledged bidirectional sync with OFF, such that you could just use Grocy for data entry and publish your changes (in addition to just mining it for information). Is this style of integration the thing that's holding the ticket open?

Could you clarify the scope of this ticket? What's your vision for an "Open Food Fact Integration", either "overall" as a pie-in-the-sky eventuality, or as a concrete implementation target that could serve as good enough for now?

**Comment by berrnd** [comment]
> I'm wondering what's the perception of "on-topic" for this ticket?
> Could you clarify the scope of this ticket?

[I already did that nearly 5 years ago](https://github.com/grocy/grocy/issues/158#issuecomment-468812212) - will never understand what's so hard on simply reading stuff and instead making the most spammy thread hear more spammy.

> _I consider to do the following two things:_
> 
>    - _Integrate the "barcode plugin lookup API only thing" into the UI_
>    - _Create a plugin for openfoodfacts.org_

Locking this thread now, since I'm tired of managing the off-topic discussion.

**Comment by berrnd** [comment]
Changelog excerpt for this:

> - Added a new product picker workflow "External barcode lookup"
>   - This executes the configured barcode lookup plugin with the given barcode
>   - If the lookup was successful, the product edit page of the created product is displayed, where the product setup can be completed (if required)
>   - After that, the transaction is continued with that product as usual
>   - A plugin for [Open Food Facts](https://world.openfoodfacts.org/) is now included and used by default (see the `config.php` option `STOCK_BARCODE_LOOKUP_PLUGIN` and maybe change it as needed)
>     - The product name and image (and of course the barcode itself) are taken over from Open Food Facts to the product being looked up
>   - => Quick video demo (using a Barcode Laser Scanner): https://www.youtube.com/watch?v=-moXPA-VvGc
>   - => Quick video demo (using Browser Camera Barcode Scanning): https://www.youtube.com/watch?v=veezFX4X1JU

Finally add new products faster than ever, without the need for other tools. 🥳

Feel free to play around with that on the [pre-release demo](https://demo-prerelease.grocy.info/).
